src/miniscript/iter.rs

```diff
@@ -0,0 +1,552 @@
+// Miniscript
+// Written in 2020 by
+//     Dr Maxim Orlovsky <orlovsky@pandoracore.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+use super::decode::Terminal;
+use super::{Miniscript, MiniscriptKey, ScriptContext};
+use std::collections::VecDeque;
+use std::ops::Deref;
+use std::sync::Arc;
+
+/// Iterator-related extensions for [Miniscript]
+impl<Pk: MiniscriptKey, Ctx: ScriptContext> Miniscript<Pk, Ctx> {
+    /// Creates a new [Iter] iterator that will iterate over all [Miniscript] items within
+    /// AST by traversing its branches. For the specific algorithm please see
+    /// [Iter::next] function.
+    pub fn iter(&self) -> Iter<Pk, Ctx> {
+        Iter::new(self)
+    }
+
+    /// Creates a new [PkIter] iterator that will iterate over all plain public keys (and not
+    /// key hash values) present in [Miniscript] items within AST by traversing all its branches.
+    /// For the specific algorithm please see [PkIter::next] function.
+    pub fn iter_pk(&self) -> PkIter<Pk, Ctx> {
+        PkIter::new(self)
+    }
+
+    /// Creates a new [PkhIter] iterator that will iterate over all public keys hashes (and not
+    /// plain public keys) present in Miniscript items within AST by traversing all its branches.
+    /// For the specific algorithm please see [PkhIter::next] function.
+    pub fn iter_pkh(&self) -> PkhIter<Pk, Ctx> {
+        PkhIter::new(self)
+    }
+
+    /// Creates a new [PkPkhIter] iterator that will iterate over all plain public keys and
+    /// key hash values present in Miniscript items within AST by traversing all its branches.
+    /// For the specific algorithm please see [PkPkhIter::next] function.
+    pub fn iter_pk_pkh(&self) -> PkPkhIter<Pk, Ctx> {
+        PkPkhIter::new(self)
+    }
+
+    /// Enumerates all child nodes of the current AST node (`self`) and returns a `Vec` referencing
+    /// them.
+    pub fn branches(&self) -> Vec<&Miniscript<Pk, Ctx>> {
```

Comment:
It appears this function is only ever used to obtain the nth child of a node. It would save some allocations and simplify the logic if you replaced it with
```
pub fn nth_child(&self) -> Option<&Miniscript<Pk, Ctx>> { ... }
```

Reply:
I've kept this fn for future possible external use but refactored the iterator code in such way that it call branches only once